internal/handlers/common/compare.go

```diff
@@ -253,3 +263,30 @@ func compareOrdered[T constraints.Ordered](a, b T) compareResult {
 func compareNumbers(a float64, b int64) compareResult {
 	return compareOrdered(a, float64(b))
 }
+
+// compareDocuments converts two documents to fjson and compares them.
+func compareDocuments(a, b *types.Document) compareResult {
+	if string(must.NotFail(fjson.Marshal(a))) == string(must.NotFail(fjson.Marshal(b))) {
+		return equal
+	}
+
+	return notEqual
+}
+
+// compareArray compares two arrays and returns document where array equals exactly the specified array
+// or array contains an element that equals the array.
+func compareArray(filterArr, docArr *types.Array) compareResult {
```

Comment:
That's a good description but a bad function name.

Let's make `compareArrays` that compares that arrays are exactly the same.

Let's then make `matchArrays` that does the "smart" thing

Reply:
This func actually is going to do the smart thing when it process cases with $gt, $lt, etc. It's just has not been implemented yet. I'm gonna add more logic to this func when I finish with simple $eq and implicit filter. Or I'm adding handler for $gt operator to this PR. WDYT?

Reply:
I think all `compareXXX` functions should not have any smarts. We almost could call them `equalXXX` except that they also check if a is less or greater than b. We need a separate function when we need to compare that documents and arrays are strictly equal. For example, we might end up deleting our custom functions that do that in tests: https://github.com/FerretDB/FerretDB/blob/main/internal/util/testutil/equal.go

MongoDB documentation calls those smarts "matching", so that could be a good prefix for functions with smarts.

Reply:
Another options is to move `compare` function for arrays and documents to types.Array and types.Document methods, but that's not possible for scalar types like int64.

Anyway, let's do all that later. Right now, we should separate simple comparision and smart comparision functions